### DISCUSSION TITLE
UI Demo Video

### DISCUSSION THREAD

**Discussion by Roland-J** [Roland-J/SmartSkillup#5] [Show and tell]
Hello,

This is a video showcasing the various ways that the SMS UI offers interactivity for its clients, hopefully providing a sleek and responsive interface.

https://user-images.githubusercontent.com/107378114/183272284-e83c17d4-d60e-457c-8fb1-6bf8c5058d78.mp4

## Addon Authors

This functionality has been made made possible via a new and user friendly "event self-registration" feature for images and texts, very similar in syntax and ease-of-use to the existing windower.register_events.
 - Previously, you would have to register a mouse handler with the mouse event, track its cryptic variables, and learn to successfully use them to determine when x scenario occurred and which variables to feed your integrated functions; all of which requires complex logic and imposes a rather large and unneeded barrier to entry for UI development.
 - With this new method, it's as simple as `image:register_event('left_click', my_function)`! Never write a mouse event handler again!!!
    - To see the background code: https://github.com/Roland-J/SmartSkillup/blob/main/libs/sms_images.lua#L452-L539
    - To see the sort of code you'd write:
        - Registration: https://github.com/Roland-J/SmartSkillup/blob/main/libs/sms_ui.lua#L338 _(Much simpler than above, right?!)_
        - Function: https://github.com/Roland-J/SmartSkillup/blob/main/libs/sms_ui.lua#L624-L645
 
### New Widgets Lib
However, I'm currently investigating merging my custom images and texts libraries into a new widgets library in order to unify the background mouse handlers so that images and texts can communicate during events. Here is a demonstration of where it's at thus far.

Keep in mind that this demo is built on less than 200 lines of code! My widgets library is jam packed with code that helps the addon authors generate UIs as easily as I can manage to help them to.

https://user-images.githubusercontent.com/107378114/184052805-a73ce2c2-3c09-431b-8068-792f42441937.mp4

### Update
I've finally managed to get the elements to stay synced across multiple drags, resizes, and addon loads! Phew! You may think it's no big deal, as SMS already does that, but SMS relies on the addon author tracking a table of offsets and informing the libs how to resize the items on resize/load, whereas I've just coded this lib to do all of that on the fly! All you give it is a default x/y pos and size, the lib handles the dynamic positioning on resizing and reloads (reload may have a custom size, after all, so it has to size the elements then, too). This should make the widgets lib much more addon-author-friendly than SMS's UI's current framework!

https://user-images.githubusercontent.com/107378114/184249200-b47d6668-9327-48f8-a2cc-8d23d36c4d6d.mp4

I hope that this functionality can make it into the v4 codebase to offer addon authors a better experience during development and make it easier for new, cool, and useful UIs to become available for the userbase! It's my understanding that something similar is already in the v5 codebase, which is great!

Check it out! Let me know if you'd like to re-use my code. Thanks!

**Comment by fixnjets** [comment]
id like to use
